Remove unconfimed email user accounts

MassiGy opened the pull request:
Hey there, 

This pull request is essentially here to setup a documentation workflow base on jsDoc, and also to setup the Database Cleanup System ( issue #109 ).

_Please go and check it out, and let me know your thoughts_

Thanks in advance
Massiles.

MassiGy commented:
This pull request is the version 1.0.0 for the Database Cleanup System #109. Please refer to the git commits and the code for more info.

Typos in git commits : **thought** not **tought**

Thanks.
Massiles.

supernova350 reviewed (changes requested):
`setInterval` will not work for this. A job scheduler like [Agenda](https://www.npmjs.com/package/agenda) with a persistence layer would be a better idea.

MassiGy commented:
Hey there !

I've just removed the `setInterval` and replaced it with the `npm/@node-sechduler` job scheduler.
I set the schedule so as the job executes on the first day of every month at 00:00 .

_Check out the git commit & code for more info._

Thanks.
Massiles.

supernova350 reviewed (approved):
LGTM! A cron job is what things like this should use.

MassiGy commented:
Hey there @supernova350,

Thanks for the great feedback. I really appreciate it.

After some research, I found that even this solutions does not solves the problem that you've pointed to. The problem is that our scheduling is an in-process scheduling, i.e. these job are scheduled but will only fire as long as our script is running.

**How can we fix that ?**

-Well, what we can do is to set our cron job not based on the script state, but on the infrastructural state. This means that we will use the server cron deamon. But, this will only work if the server is using linux and that we have only one server. Otherwise, it will be very difficult to set this up.

-Another way around, is to use a database/store, to have a persistent scheduling, like `npm/@agenda` with `mongoDB` or `npm/@bull` with `redis`. With that, we can remember that last time the job is executed.

-Also, what we can do is to create a row or a table in our database to store related information about our job, and recreate that behavior by ourselves.


**Is it worth it ?**

Well, with the first solution definitely not !. The reason is that behavior will make us dependent on our infrastructure, and we do not know it, since we do not own the server in which our code is running.

Some disadvantages are shown here, source on this [link](medium.com/geekculture/cron-jobs-in-node-js-8df170445588)

![Screenshot from 2022-03-26 20-27-19](https://user-images.githubusercontent.com/82173113/160254290-61cb97ca-ca0f-459e-a057-a20fcfa61c83.png)
---
So the remaining way to go is the second and the third, I'am joining them since they are very similar, the whole point is to set a database/store to memorize the last time our job executed.


**My thoughts on the problem**
I really understand the problem, but I do not think that this is big enough to go and spend development time creating a solution for it. The reason is, even if we redeploy and the scheduling restarts, the worst thing that will happen it that some users can still access our platform for one more month before we delete their account from our database.



This is my thoughts on the current problem, this does not mean that we do not need to solve it, but this means that it is not urgent.

**Ideas from this**
-We need to build a system that will remind the user to confirm their email address before deleting their account on the database.



_Please let me know your thoughts on this, and if something is wrong._

I look forward to hearing from you.
Thanks in advance.
Massiles.

supernova350 commented:
What you said is correct, but is too much for a project of this scale. The simpler the better. 

Let's focus on making the project go live before thinking about the size of the database. If it is extremely important to you, I'd recommend storing a `created_at` timestamp for each user account and use that in an sql script to manually purge inactive accounts (ex. if current - created > 30d, purge).

MassiGy commented:
hey @supernova350, 

Thanks for your great feedback as always.

Yeah, I totally agree with you, and I want to say that that's is what I am currently developing, in other words this pull request is just to set an alpha version of the system. The current version is basically fetching the users with `email_confirmed` set to `false` and purge them.
The next version, that I am currently developing, will try to use sequelize to retrieve the age of each row in our user table. Which in fact is the same thing as what you said. I just need to search a bit, since I am an express/mongodb developer and I never dealt with sequelize before XD.

_This is my current plan, please tell me if something is wrong._

Thanks in advance.
Massiles.

MassiGy commented:
Hey @supernova350 ,

Guess what, I've just finished it and the trick was really simple XD. Please let me know what you think about it.

Thanks in advance.
Massiles.

supernova350 commented on src/background_jobs/unconfirmed_emails.js:
```diff
@@ -25,22 +24,23 @@ module.exports.remove_unconfirmed_email_users = async () => {
         const ids = [];
         found_users.forEach((element) => {
-            ids.push(element.id)
+            if (Date.parse(element.createdAt) < (Date.now() - 30 * 24 * 60 * 60 * 1000)) {
```
This should be a constant. It's always a good idea to not hardcode values like these, as you might want to change them later.

MassiGy replied:
Hey @supernova350,

I just made it like that ( not a constant ), because first of all, this should be calculated at the time of the execution. Also, I did not use it anywhere else. Just a personal preference.

_What you think ?_

Thanks in advance.
Massiles.

supernova350 replied:
Well, calculating `24 * 60 * 60...` does not need to be done at runtime. If you have a constant, for example, `const ACCOUNT_MAX_UNCONFIRMED_AGE_DAYS = 30;`, it just makes the the numbers easier to understand. No problem though, just a suggestion.

supernova350 commented:
Otherwise, looks good! The cron job will do its work, but if the process is not running, it will not execute the method. It may help to make the cron interval 1d, so there's a better chance of the method being called.

MassiGy commented:
hey @supernova350,

About the cron interval we made it more then 20 days for better UX. In other words, if a user did not confirmed his email address within the first 30 days, we can be sure at 99% that he wont do it after that. But, we can not affirm that after just 1 day.

Also, it is more of a convention to do it like that.

Hope that is clear.
Massiles.

supernova350 commented:
The cron interval does not really matter, since we're checking if the account is unconfirmed AND older than 30 days. Theoretically, we could have an async while(true) and check unconfirmed accounts there. However, that would be inefficient. The interval is just to reduce load on the server.

MassiGy commented:
Exactly, the cron is just to avoid checking for unwanted users at every second. Which is really inefficient. Also the cron schedule can be really helpful since it is a way of planning events and job, we can use it to create a chronological sequence of those. For example, we can use it to create the system that I've suggested, which will remind the user to confirm his/her email address.

Massiles.

supernova350 commented:
> About the cron interval we made it more then 20 days for better UX.
So, this doesn't really make sense. No problem though!

MassiGy commented:
Well, yes it does !
We do not want to remove user accounts everyday as you suggested.

 Imagine that we do that and a user registers today, if he do not confim his email address today, then his account will be purged  tomorrow. Which will resault in a worse UX.

Hope that is clear.
Massiles

MassiGy commented:
> > About the cron interval we made it more then 20 days for better UX.
> 
> So, this doesn't really make sense. No problem though!

Yeah, I see what you mean.
I should of said the `unconfirmed_user_account_max_age` instead of `cron`.

The `cron` is used to lower the server load, and the  `unconfirmed_user_account_max_age= 30 days`   is for better UX.

Hope that is clear.
Massiles.

supernova350 commented:
That’s what I was trying to say. My bad if I was unclear.

MassiGy commented:
Well I think with that been said, we can agree that this first version of the system is on point.

_Aren't you?_

Massiles

supernova350 commented:
I agree. It should be ready.

William-McGonagle commented:
Looks good- Thanks @MassiGy @supernova350!

MassiGy commented:
Thanks  for the great feedback as always.

Massiles.
